docs: add documentation for modules in contributing

DevelopmentCats opened the pull request:
Let me know if this align's with what you were looking for in this modules doc. @matifali

DevelopmentCats committed:
docs: add documentation for Coder modules and their usage

DevelopmentCats committed:
docs: add guide for module development in manifest

DevelopmentCats committed:
docs: update module structure formatting in contributing guide

DevelopmentCats committed:
docs: add optional installation script and enhance resource configuration in module guide

matifali commented:
Can you add a diagram describing how a module fits into Cider architecture?

Also I would suggest being lean on the module usage section i.e. example uses, and more detailed on how to write the module and what is the significance of each type of resource an author can add to the module. What are those sources provided. 

Here we can also consider examples like what will be needed to write a module to:

- Enable an external ide integration.
- Enable an externality auth with a service by installing CLI binary ir injecting some env variable
- Writing some configuration files.
- Install a VSCode extension? 


In these we can also link existing matching module implementations as examples.

DevelopmentCats committed:
docs: enhance module contribution guide with architecture overview and detailed resource explanations

DevelopmentCats committed:
docs: improve link visibility in architecture overview with styled arrows

DevelopmentCats committed:
docs(modules): linting

bpmct commented on docs/about/contributing/modules.md:
```diff
@@ -0,0 +1,449 @@
+
+### External IDE Integration
+
+**Browser-based IDE Pattern**: Install web-based development environment with health monitoring and configuration management.
```
```suggestion
**Browser-based IDE**: Install web-based development environment with health monitoring and configuration management.
```

matifali commented on docs/about/contributing/modules.md:
```diff
@@ -0,0 +1,449 @@
+
+**Key properties**:
+
+- `agent_id`: The Coder agent
+- `external`: `true` for protocol handlers, `false` for web apps
+- `healthcheck`: Monitor service availability
+- `subdomain`: Access method for web apps
+- `url`: Service URL or protocol handler
```
`order` and `group` are also important to tell her. Maybe not if we link provider docs.

matifali commented on docs/about/contributing/modules.md:
```diff
@@ -0,0 +1,449 @@
+# Provides: name, id, access_url, start_count
+
+data "coder_workspace_owner" "me" {}
+# Provides: name, email, full_name
```
```suggestion
# Provides: name, email, full_name, groups of the workspace owner. 
```

matifali commented on docs/about/contributing/modules.md:
````diff
@@ -0,0 +1,449 @@
+}
+```
+
+## Testing and Documentation
+
+### Module Testing
+
+While not mandatory, **tests are recommended** for all modules. A `main.test.ts` file validates that the module works correctly. Tests use **Bun** as the test runner and verify that:
+
+- **Custom configurations** work with different variable values
+- **Default configuration** creates expected resources
+- **Input validation** catches invalid parameters
+- **Required variables** are properly validated
+
+The test framework provides utilities from `~test` import to:
+
+- Initialize Terraform in the module directory
+- Apply Terraform with test variables
+- Validate the resulting infrastructure state
+- Test error conditions and edge cases
+
+Tests run in Docker with `--network=host` capabilities and help ensure module quality and reliability.
+
+### Documentation Requirements
+
+Every module requires a `README.md` with YAML frontmatter containing metadata like display name, description, icon, maintainer, and tags. The README must include:
+
+- **Purpose and functionality** - What the module does and when to use it
+- **Variable documentation** - All input variables with types, defaults, and descriptions  
+- **Resource listing** - What Coder resources the module creates
+- **Usage examples** - Basic and advanced configuration examples
+- **Requirements** - Any prerequisites or dependencies
+
+The documentation format is validated automatically and must pass validation before modules can be accepted. See the [Registry Contributing Guide](https://github.com/coder/registry/blob/main/CONTRIBUTING.md) for detailed formatting requirements and examples.
````
I think this whole section can be removed or moved to Contributing docs in registry repo.

matifali commented on docs/about/contributing/modules.md:
```diff
@@ -0,0 +1,449 @@
+
+## Common Use Case Patterns
+
+### External IDE Integration
+
+**Browser-based IDE Pattern**: Install web-based development environment with health monitoring and configuration management.
+
+**Desktop IDE Pattern**: Provide protocol-based integration for native applications with backend service management.
+
+**Editor Extension Pattern**: Configure existing editors with workspace-specific settings and extensions.
+
+### Authentication and Secrets
+
+**OAuth Integration Pattern**: Handle external authentication flows with token management and service configuration.
+
+**API Key Management Pattern**: Securely distribute credentials and configure authenticated tools.
+
+**Certificate Management Pattern**: Handle SSL certificates and secure communication setup.
+
+### Development Environment Setup
+
+**Language Runtime Pattern**: Install and configure programming language environments with package managers.
+
+**Database Service Pattern**: Deploy and configure database services with connection management.
+
+**Configuration Management Pattern**: Apply dotfiles, settings, and workspace personalization.
+
```
Can we also link example modules heres?

matifali commented on docs/about/contributing/modules.md:
```diff
@@ -0,0 +1,449 @@
+
+**Configuration Management Pattern**: Apply dotfiles, settings, and workspace personalization.
+
+## Contributing to the Registry
+
+The [Coder Registry](https://github.com/coder/registry) hosts all community and official modules. For detailed contribution guidelines, code standards, and submission processes, refer to the [Registry Contributing Guide](https://github.com/coder/registry/blob/main/CONTRIBUTING.md).
+
+**Quick contribution steps**:
+
+1. Fork the [registry repository](https://github.com/coder/registry)
+2. Follow the [registry contribution guidelines](https://github.com/coder/registry/blob/main/CONTRIBUTING.md) for module creation process
+3. Implement tests: `bun test`
+4. Format code: `bun run fmt`
+5. Validate documentation: `bun run readme-validate`
+6. Submit pull request with detailed description
+
+**Important**: The registry repository contains specific requirements for module structure, testing, documentation format, and submission process. Always consult the [official contributing documentation](https://github.com/coder/registry/blob/main/CONTRIBUTING.md) for the most current guidelines.
+
```
Again move this to regstry repo and link.

DevelopmentCats committed:
docs(modules): expand module types section with examples for IDE, Helper, Desktop, AI, Integration, and Web-based Tools and remove duplicate explanation from end of doc.

DevelopmentCats committed:
Update docs/about/contributing/modules.md

Co-authored-by: Atif Ali <atif@coder.com>

DevelopmentCats committed:
docs(modules): remove testing and documentation sections

DevelopmentCats committed:
docs(modules): remove contributing guide and lint

DevelopmentCats committed:
docs(modules): enhance module list with hyperlinks for better navigation

DevelopmentCats commented on docs/about/contributing/modules.md:
```diff
@@ -0,0 +1,449 @@
+
+**Key properties**:
+
+- `agent_id`: The Coder agent to execute the script on
+- `display_name`: Name shown in the Coder dashboard during execution
+- `run_on_start`: Execute when workspace starts (most common)
+- `run_on_stop`: Execute during workspace shutdown for cleanup
+- `script`: The actual shell script content to execute
+- `start_blocks_login`: Whether script completion is required before user access
```
I like the idea of listing the stuff here because the provider doc's feel like completely external documentation and it took me longer than I would like to admit to realize the docs were all available here when I started working on modules.

DevelopmentCats committed:
docs(modules): update contributing guide with prerequisites and enhanced resource links for better clarity

addressed all comments up to commit time aside from unresolved comments awaiting replies

matifali reviewed (commented):
I did another pass and it looks way better now. Please address comments.

DevelopmentCats committed:
docs: update description

Co-authored-by: Atif Ali <atif@coder.com>

DevelopmentCats committed:
docs: remove depreciated vault-secrets

Co-authored-by: Atif Ali <atif@coder.com>

DevelopmentCats committed:
docs: update workspace execution content

Co-authored-by: Atif Ali <atif@coder.com>

DevelopmentCats committed:
docs(modules): refine IDE modules section and categorize browser-based and desktop IDEs for clarity

DevelopmentCats committed:
docs(modules): clarify browser-based and desktop IDEs functionality in contributing guide

DevelopmentCats committed:
docs(modules): replace mermaid code with png and enhance parameter and workspace enhancement modules sections

DevelopmentCats committed:
docs(modules): move header links to body

DevelopmentCats committed:
docs(modules): fix image path for module architecture diagram

matifali commented on docs/about/contributing/modules.md:
````diff
@@ -0,0 +1,406 @@
+}
+```
+
+> **⚠️ Important**: Health check URLs are service-specific. Common paths include `/health`, `/healthz`, `/ping`, `/status`, or `/api/health`. Check your service's documentation or use the main service URL if no dedicated health endpoint exists.
````
Use GFM style

matifali commented on docs/about/contributing/modules.md:
````diff
@@ -0,0 +1,406 @@
+Example:
+
+```tf
+resource "coder_env" "config" {
````
Use a different example. Something like `VAUKT_ADDR` or `VAULT_TOKEN`

matifali commented on docs/about/contributing/modules.md:
````diff
@@ -0,0 +1,406 @@
+# Provides: name, email, full_name, groups of the workspace owner.
+```
+
+### Variable Design
````
Why is it called design? Maybe we call it Module Inputs variable types. Also this is a terraform concept and not Coder so should also be able to link to some upstream docs.

matifali commented on docs/about/contributing/modules.md:
````diff
@@ -0,0 +1,406 @@
+}
+```
+
+### Script Development
````
Same comment. I think we should drop the words Development and use an action sentence like "Running Scripts"
